src/qmc_states.jl

```diff
@@ -0,0 +1,332 @@
+
+    # Set up spectral_states
+    nreplicas = num_replicas(replica_strategy)
+    if nreplicas > 1
+        spectral_states = ntuple(nreplicas) do i
+            SpectralState(
+                (SingleState(
+                    hamiltonian,
+                    deepcopy(v),
+                    deepcopy(wm),
+                    deepcopy(shift_strategy),
+                    deepcopy(time_step_strategy),
+                    shift,
+                    time_step,
+                    "_$i"),
+                ),
+                spectral_strategy
+            )
+        end
+    else
+        spectral_states = (SpectralState(
+            (SingleState(hamiltonian, v, wm, shift_strategy, time_step_strategy, shift, time_step),),
+            spectral_strategy
+        ),)
+    end
+
+    return ReplicaState(
+        hamiltonian, spectral_states, Ref(Int(maxlength)),
+        Ref(simulation_plan.starting_step), # step
+        simulation_plan,
+        # Ref(Int(laststep)),
+        reporting_strategy, post_step_strategy, replica_strategy
+    )
+end
+
+num_replicas(::ReplicaState{N}) where {N} = N
+num_spectral_states(::ReplicaState{<:Any, S}) where {S} = S
+
+function Base.show(io::IO, st::ReplicaState)
+    print(io, "ReplicaState")
+    print(io, " with ", num_replicas(st), " replicas  and ")
+    print(io, num_spectral_states(st), " spectral states")
+    print(io, "\n  H:    ", st.hamiltonian)
+    print(io, "\n  step: ", st.step[], " / ", st.simulation_plan.last_step)
+    print(io, "\n  replicas: ")
+    for (i, r) in enumerate(st.spectral_states)
+        print(io, "\n    $i: ", r)
+    end
+end
+
+"""
+    state_vectors(state::ReplicaState)
+Returns a collection of configuration vectors from the `state`.
+The vectors can be accessed by indexing the resulting collection, where the row index
+corresponds to the spectral state index and the column index corresponds to the row index.
+
+See also [`single_states`](@ref), [`SingleState`](@ref), [`ReplicaState`](@ref),
+[`SpectralState`](@ref), [`PMCSimulation`](@ref).
+"""
+@inline function state_vectors(state::ReplicaState{N,S}) where {N,S}
+    # Annoyingly this function is allocating if N > 1
+    return SMatrix{S,N}(
+        state.spectral_states[fld1(i,S)].single_states[mod1(i,S)].v for i in 1:N*S
+    )
+    # return SMatrix{S,N}(mapreduce(state_vectors, hcat, state.spectral_states))
+end
+
+"""
+    single_states(state::ReplicaState)
+Returns a collection of `SingleState`s from the `state`.
+The `SingleState`s can be accessed by indexing the resulting collection, where the row index
+corresponds to the spectral state index and the column index corresponds to the row index.
+
+See also [`state_vectors`](@ref), [`SingleState`](@ref), [`ReplicaState`](@ref),
+[`SpectralState`](@ref), [`PMCSimulation`](@ref).
+"""
+function single_states(state::ReplicaState{N,S}) where {N,S}
+    return SMatrix{S,N}(mapreduce(single_states, hcat, state.spectral_states))
+end
```

Comment:
Would it make sense to replace `single_states` with `getindex` such that
```julia
single_states(state)[i, j]
```
would become
```
state[i][j]
```